[Merged by Bors] - Cleanups in diagnostics

mockersf opened the pull request:
- changed `EntityCountDiagnosticsPlugin` to not use an exclusive system to get its entity count
- removed mention of `WgpuResourceDiagnosticsPlugin` in example `log_diagnostics` as it doesn't exist anymore
- added ability to enable, disable ~~or toggle~~ a diagnostic (fix #3767)
- made diagnostic values lazy, so they are only computed if the diagnostic is enabled
- do not log an average for diagnostics with only one value
- removed `sum` function from diagnostic as it isn't really useful
- ~~do not keep an average of the FPS diagnostic. it is already an average on the last 20 frames, so the average FPS was an average of the last 20 frames over the last 20 frames~~
- do not compute the FPS value as an average over the last 20 frames but give the actual "instant FPS"
- updated log format to use variable capture
- added some doc
- the frame counter diagnostic value can be reseted to 0

DJMcNab reviewed (commented):
The changes related to `max_history_length in 0..=1` also strike me as odd, since there's no reason you can't have an average of a single item. The intention to not print the average of an item with no history makes sense, but I don't think this quite encodes the logic there properly.

The methods on `Diagnostic` and `Diagnostics` could also do with some docs whilst we're modifying this (I haven't checked if there's already a pr).

mockersf commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -129,40 +129,97 @@ impl Diagnostic {
     // This uses a [`StableHashMap`] to ensure that the iteration order is deterministic between
     // runs when all diagnostics are inserted in the same order.
-    diagnostics: StableHashMap<DiagnosticId, Diagnostic>,
+    diagnostics: StableHashMap<(DiagnosticId, DiagnosticState), Diagnostic>,
```
I fixed the `add`, `disable`, `add` scenario by removing any potential `Disabled` diagnostic before re-adding it...

Yup, it's to avoid the cost of the extra check and... the fact that the code became not pretty by adding two layers of `if`
```rust
    // with the state in the key
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        self.diagnostics.get(&(id, DiagnosticState::Enabled))
    }

    // with the state as an property
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        if let Some(diagnostic) = self.diagnostics.get(&id) {
            if diagnostic.state == DiagnosticState::Enabled {
                return Some(diagnostic);
            }
        }
        None
    }
```

As I write this, it just came to mind that it would have work with
```rust
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        self.diagnostics
            .get(&id)
            .filter(|diagnostic| diagnostic == DiagnosticState::Enabled)
    }
```

Still I find the state as key to be prettier, but I'm open to change

DJMcNab replied:
That filter could probably also be `Diagnostic::is_enabled`

I find that this is a messy implementation, especially given that it makes the `state` method `O(n)`. If I'd found this during one of my archaeology sessions, I'd add it to my refactorings list.

Now maybe the performance effects are worthwhile, but I'm exceedingly skeptical.

mockersf replied:
you convinced me 👍

mockersf commented on crates/bevy_diagnostic/src/frame_time_diagnostics_plugin.rs:
```diff
@@ -37,14 +37,17 @@ impl FrameTimeDiagnosticsPlugin {
-        state.frame_count += 1.0;
-        diagnostics.add_measurement(Self::FRAME_COUNT, state.frame_count);
+        diagnostics.add_measurement(Self::FRAME_COUNT, || {
+            state.frame_count += 1.0;
```
It depends what is it supposed to measure: the absolute frame count, or the frame count while the diagnostic is enabled? I would be for the later

mockersf commented:
> The changes related to `max_history_length in 0..=1` also strike me as odd, since there's no reason you can't have an average of a single item. The intention to not print the average of an item with no history makes sense, but I don't think this quite encodes the logic there properly.

It was me being lazy and not wanting to do it in the logger... I changed that 👍 

> The methods on `Diagnostic` and `Diagnostics` could also do with some docs whilst we're modifying this (I haven't checked if there's already a pr).

Added some doc 👍

mockersf commented on crates/bevy_diagnostic/src/log_diagnostics_plugin.rs:
```diff
@@ -62,17 +62,14 @@ impl LogDiagnosticsPlugin {
                     // The ) hugging the value is more aesthetically pleasing
                     "{name:<name_width$}: {value:>11.6}{suffix:1} (avg {average:>.6}{suffix:})",
                     name = diagnostic.name,
-                    value = value,
```
just to make sure that Bevy MSRV is latest!

alice-i-cecile commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -40,19 +40,26 @@ pub struct Diagnostic {
 }
 
 impl Diagnostic {
+    /// Add a new value measured now.
```
The meaning of "now" is very unclear from reading the doc string alone.

alice-i-cecile commented on crates/bevy_asset/src/diagnostic/asset_count_diagnostics_plugin.rs:
```diff
@@ -48,6 +48,6 @@ impl<T: Asset> AssetCountDiagnosticsPlugin<T> {
 
     pub fn diagnostic_system(mut diagnostics: ResMut<Diagnostics>, assets: Res<Assets<T>>) {
-        diagnostics.add_measurement(Self::diagnostic_id(), assets.len() as f64);
+        diagnostics.add_measurement(Self::diagnostic_id(), || assets.len() as f64);
```
Why do we need this zero-argument closure now?

mockersf replied:
to be able to not compute the value if the diagnostic is not enabled

alice-i-cecile commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -40,19 +40,26 @@ pub struct Diagnostic {
 
 impl Diagnostic {
+    /// Add a new value measured now.
     pub fn add_measurement(&mut self, value: f64) {
```
I am really not thrilled with the fact that all measurements are assumed to be f64, but that's probably out of scope.

mockersf replied:
yup me neither 🙁

mockersf commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -77,19 +84,18 @@ impl Diagnostic {
     }
 
+    /// Get the oldest value from this diagnostic.
     pub fn value(&self) -> Option<f64> {
```
I agree, I have another PR that changes that: #3056

mockersf commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -40,19 +40,26 @@ pub struct Diagnostic {
 }
 
 impl Diagnostic {
```
there is, it's the `measurements` function

alice-i-cecile replied:
Ah, okay.

mockersf commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -127,42 +136,121 @@ impl Diagnostic {
+    }
+
+    /// Toggle a [`Diagnostic`] by its [`DiagnosticId`], and returns the new state.
+    pub fn toggle(&mut self, diagnostic_id: DiagnosticId) -> Option<DiagnosticState> {
```
removed!

alice-i-cecile reviewed (approved):
My concerns have all been addressed; this is an important bit of cleanup.

alice-i-cecile reviewed (approved):
bors r+

CAD97 commented on crates/bevy_diagnostic/src/diagnostic.rs:
```diff
@@ -37,22 +37,30 @@ pub struct Diagnostic {
+        if self.max_history_length > 1 {
+            if self.history.len() == self.max_history_length {
+                if let Some(removed_diagnostic) = self.history.pop_front() {
+                    self.sum -= removed_diagnostic.value;
```
A potential solution would be to switch form storing the sum to storing the simple moving average directly. An incremental SMA is calculated as

$$ SMA_t = \frac{ X_t + SMA_{t-1} \times (n-1) }{ n } $$

where the impact of any rounding error trends towards zero.

The main advantage of the incremental SMA over the typical and current $\frac{ X_t + X_{t-1} + \cdots + X_{t-n} }{ n }$ is not needing to store the historical readings, though, and it's perhaps useful to have the history, especially for discrete diagnostics rather than continuous.

Perhaps the best solution really is to separate discrete from continuous diagnostics.

CAD97 replied:
Strike that formula that's what I get for going off of memory. That's not a correct simple moving average; the SMA does require keeping track of the history. The correct incremental SMA formula is $$SMA_{k,\text{next}} = SMA_{k,\text{prev}} + \frac{1}{k}\left( p_{n+1} - p_{n-k+1} \right)$$. What I wrote above is some odd form of a weighted moving average where each sample has weight IIUC $1/(n+\text{age})$.
